src/withdrawals/tasks.py

```diff
@@ -0,0 +1,351 @@
+            and last_withdrawals_block + partial_withdrawals_blocks_interval >= block_number
+        ):
+            return False
+        return True
+
+    async def _fetch_last_withdrawals_block(
+        self, vault_address: ChecksumAddress, from_block: BlockNumber
+    ) -> BlockNumber | None:
+        """
+        Fetch withdrawal events for the required interval.
+        Validate that the withdrawal was successful via a consensus request.
+        Return the last event block number.
+        """
+        vault_contract = VaultContract(vault_address)
+        events = await vault_contract.get_validator_withdrawal_submitted_events(from_block)
+        consensus_validators = await fetch_consensus_validators(
+            [event.public_key for event in events]
+        )
+        public_key_to_index = {val.public_key: val.index for val in consensus_validators}
+
+        for event in events:
+            slot = await calc_slot_by_block_number(event.block_number)
+            pending_partial_withdrawals = await consensus_client.get_pending_partial_withdrawals(
+                slot
+            )
+            for withdrawal in pending_partial_withdrawals:
+                if (
+                    int(withdrawal['validator_index']) == public_key_to_index[event.public_key]
+                    and Web3.to_wei(withdrawal['amount'], 'gwei') == event.amount
+                ):
+                    return event.block_number
+
+        return None
+
+
+async def _get_withdrawals_data(
+    chain_head: ChainHead,
+    queued_assets: Gwei,
+    consensus_validators: list[ConsensusValidator],
+    validator_min_active_epochs: int,
+    oracle_exit_indexes: set[int],
+) -> dict[HexStr, Gwei]:
+    # filter active validators
+    partial_withdrawable_validators = [
+        val
+        for val in consensus_validators
+        if _is_partial_withdrawable_validator(val, chain_head.epoch)
+    ]
+    available_partial_withdrawals_capacity = sum(
+        val.balance - MIN_ACTIVATION_BALANCE_GWEI for val in partial_withdrawable_validators
+    )
+    if available_partial_withdrawals_capacity >= queued_assets:
+        return _get_partial_withdrawals_data(
+            validator_balances={
+                val.public_key: val.balance for val in partial_withdrawable_validators
+            },
+            queued_assets=queued_assets,
+        )
+    if settings.disable_full_withdrawals:
+        logger.info('Available partial withdrawals capacity is less than queued assets')
+        return {}
+
+    min_activation_epoch = max(0, chain_head.epoch - validator_min_active_epochs)
+
+    can_be_exited_validators = _filter_exitable_validators(
+        consensus_validators=consensus_validators,
+        min_activation_epoch=min_activation_epoch,
+        oracle_exit_indexes=oracle_exit_indexes,
+    )
+    withdrawals_data: dict[HexStr, Gwei] = {}
+
+    for validator in can_be_exited_validators:
+        if sum(withdrawals_data.values()) >= queued_assets:
```

Comment:
Instead of summing values, I suggest having a counter like `total_exited_assets`